src/main/java/org/truffleruby/core/rubinius/ByteArrayNodes.java

```diff
@@ -103,6 +120,17 @@ public int size(DynamicObject bytes) {
 
     }
 
+    @CoreMethod(names = "length=", required = 1)
+    public abstract static class SetLengthNode extends CoreMethodArrayArgumentsNode {
+
+        @Specialization
+        public int setLength(DynamicObject bytes, int length) {
+            Layouts.BYTE_ARRAY.getBytes(bytes).setLength(length);
```

Comment:
Adding an assertion that the new length is <= `IOBUFFER_SIZE` could save you some annoying debugging later on. Setting the length does not resize the underlying byte array. But it's possible all paths that would lead you here would blow up on `System.arraycopy` first.

Reply:
The `length` field is totally unrelated to the length of the `byte[]` as far as I can tell for operations with IO::InternalBuffer, I didn't change that.
Maybe we don't even need to set it, or conversely we can use `append()` and higher level APIs than arraycopy. I would like to investigate this, but in another PR. This PR intends to change nothing re the ByteArrayBuilder.

Reply:
Hmm... okay. I thought the idea was the underlying `byte[]` could have excess capacity and the `length` field tracks how much of that we're actually using. Much like what `ByteList` used to do. That way you don't need to keep reallocating everytime you append a single byte.

Reply:
I think that was the original idea but there are various weird usages.
`setLength()` should not exist if it's only high-level `append`-like operations.